In [1]:
import numpy as np
import pandas as pd

In [2]:
def nor_minmax(x):
    result = (x - min(x)) / (max(x) - min(x))
    return result

In [11]:
ser = pd.Series([2, 4, 5, 6, 8])

In [12]:
nor_minmax(ser)

0    0.000000
1    0.333333
2    0.500000
3    0.666667
4    1.000000
dtype: float64

In [13]:
def nor_std(x):
    result = (x - x.mean()) / x.std()
    return result

In [14]:
nor_std(ser)

0   -1.341641
1   -0.447214
2    0.000000
3    0.447214
4    1.341641
dtype: float64

In [9]:
df = pd.read_csv("diamonds.csv")
df.head(2)

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31

In [7]:
df["carat_nor"] = nor_minmax(df["carat"])
df.head(2)

carat      cut color clarity  depth  table  price     x     y     z  \
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43   
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31   

   carat_nor  
0   0.006237  
1   0.002079

In [15]:
df = pd.read_csv("bike.csv")
df.head(1)

datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 00:00:00       1        0           0        1  9.84  14.395   

   humidity  windspeed  casual  registered  count  
0        81        0.0       3          13     16

In [16]:
df["temp"].skew()

0.003690844422472008

In [17]:
df["temp"].kurt()

-0.9145302637630794

In [18]:
df2 = df.set_index("season")
df2.head(2)

datetime  holiday  workingday  weather  temp   atemp  \
season                                                                    
1       2011-01-01 00:00:00        0           0        1  9.84  14.395   
1       2011-01-01 01:00:00        0           0        1  9.02  13.635   

        humidity  windspeed  casual  registered  count  
season                                                  
1             81        0.0       3          13     16  
1             80        0.0       8          32     40

In [19]:
df2.kurt(level="season")

holiday  workingday   weather      temp     atemp  humidity  \
season                                                                  
1       32.921622   -1.400471  0.384670  0.114545 -0.106134 -0.433792   
2       52.052753   -1.302876  0.347588 -0.232492 -0.060169 -1.007305   
3       23.550413   -1.441438  0.972714  0.175452  2.439905 -0.746044   
4       23.560818   -1.440570  0.028475 -0.493403 -0.580080 -0.961572   

        windspeed     casual  registered     count  
season                                              
1        0.335609  34.286416    4.709668  4.314757  
2        0.237428   5.527495    1.707194  0.425213  
3        1.756376   3.142335    2.093644  0.699383  
4       -0.072296  12.343876    2.310093  1.273485

In [20]:
df2["temp"].kurt(level="season")

season
1    0.114545
2   -0.232492
3    0.175452
4   -0.493403
Name: temp, dtype: float64

In [21]:
from scipy.stats import pearsonr

In [23]:
df = pd.read_excel("iris_xlsx.xlsx")
df.head(2)

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width Species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa

In [24]:
pearsonr(df["Sepal.Length"], df["Sepal.Width"])

(-0.11756978413300204, 0.15189826071144918)

In [25]:
stat, p = pearsonr(df["Sepal.Length"], df["Sepal.Width"])
print(round(stat, 3)) # 상관계수
print(round(p, 3)) # p-value

-0.118
0.152


In [26]:
result = pearsonr(df["Sepal.Length"], df["Sepal.Width"])
result

(-0.11756978413300204, 0.15189826071144918)

In [27]:
aa = result[0]
bb = result[1]
aa

-0.11756978413300204

In [28]:
df.corr() # 상관 계수 행렬

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width
Sepal.Length      1.000000    -0.117570      0.871754     0.817941
Sepal.Width      -0.117570     1.000000     -0.428440    -0.366126
Petal.Length      0.871754    -0.428440      1.000000     0.962865
Petal.Width       0.817941    -0.366126      0.962865     1.000000

상관계수를 기술하고 귀무가설 기각여부를 O 또는 X로 표기하시오.

In [29]:
from scipy.stats import bartlett

In [31]:
bartlett(df["Sepal.Length"], df["Sepal.Width"])

BartlettResult(statistic=57.372797334557305, pvalue=3.6055975610903356e-14)

두 집단의 등분산 검정 결과(bartlett)를 보았을 때 p-value가 $3.606 \times 10^{-14}$ 으로 유의수준($\alpha$, 0.05)보다 작기 때문에 두 집단의 분산이 같다는 귀무가설을 기각하고 대립가설을 채택한다.

즉, 두 집단간 변수의 분산은 다르다.

In [33]:
from scipy.stats import chi2_contingency

In [35]:
df = pd.read_csv("traffic_accident.txt", sep = "\t")
df.head(2)

seq sex serious_yn  age
0    1   M          N    3
1    2   M          N    6

In [37]:
chi2_contingency(pd.crosstab(df["serious_yn"], df["sex"]),
                 correction=False)

(4.248074934656112,
 0.03929484927552679,
 1,
 array([[286.99343955, 679.00656045],
        [ 30.00656045,  70.99343955]]))

In [38]:
from scipy.stats import ttest_1samp
from scipy.stats import ttest_rel
from scipy.stats import ttest_ind

In [39]:
df = pd.read_excel("iris_xlsx.xlsx")
df.head(2)

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width Species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa

In [42]:
df["Sepal.Length"].mean()

5.843333333333335

In [53]:
stat, p = ttest_1samp(df["Sepal.Length"], popmean = 4.8) # popmean <- 모평균
print(round(stat, 4)) # 검정통계량 t
print(round(p, 4)) # p-value 는 검정통계량 t와 t분포 기반으로 산출

15.4313
0.0


In [54]:
from statsmodels.formula.api import ols
from statsmodels.stats.anova import anova_lm

In [56]:
df = pd.read_csv("diamonds.csv")
df.head(2)

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31

In [58]:
formula = "price ~ C(cut)" # 여기서 C() 는 범주형 변수를 의미
lm = ols(formula, df).fit()
anova_lm(lm) # ANOVA table

df        sum_sq       mean_sq           F         PR(>F)
cut           4.0  1.104175e+10  2.760436e+09  175.688717  8.428307e-150
Residual  53935.0  8.474314e+11  1.571209e+07         NaN            NaN

p-value가 $8.42 \times 10^-150$으로 유의수준 5%(0.05)보다 작기 때문에 모든 집단의 평균이 같다는 귀무가설을 기각하고 대립가설을 채택.

In [63]:
df.groupby("cut")["price"].mean().reset_index()

cut        price
0       Fair  4358.757764
1       Good  3928.864452
2      Ideal  3457.541970
3    Premium  4584.257704
4  Very Good  3981.759891

In [60]:
bike = pd.read_csv("bike.csv")

In [61]:
lm_bike = ols("registered ~ season", bike).fit()
anova_lm(lm_bike)

df        sum_sq       mean_sq           F        PR(>F)
season        1.0  6.679598e+06  6.679598e+06  300.866825  1.665411e-66
Residual  10884.0  2.416376e+08  2.220118e+04         NaN           NaN

In [62]:
lm_bike = ols("registered ~ C(season)", bike).fit()
anova_lm(lm_bike)

df        sum_sq       mean_sq           F         PR(>F)
C(season)      3.0  1.099019e+07  3.663398e+06  167.975391  1.888299e-106
Residual   10882.0  2.373270e+08  2.180914e+04         NaN            NaN

In [64]:
from statsmodels.stats.multicomp import pairwise_tukeyhsd

In [65]:
posthoc = pairwise_tukeyhsd(df["price"], df["cut"])
print(posthoc) # reject는 두 집단의 평균이 같다는 귀무가설 기각 여부

     Multiple Comparison of Means - Tukey HSD, FWER=0.05      
 group1   group2   meandiff p-adj    lower      upper   reject
--------------------------------------------------------------
   Fair      Good -429.8933 0.0015  -740.4567 -119.3299   True
   Fair     Ideal -901.2158  0.001 -1180.5785 -621.8531   True
   Fair   Premium  225.4999  0.195   -59.2739  510.2738  False
   Fair Very Good -376.9979 0.0031  -663.8694  -90.1263   True
   Good     Ideal -471.3225  0.001   -642.367 -300.2779   True
   Good   Premium  655.3933  0.001   475.6466  835.1399   True
   Good Very Good   52.8954    0.9  -130.1565  235.9474  False
  Ideal   Premium 1126.7157  0.001  1008.8058 1244.6257   True
  Ideal Very Good  524.2179  0.001   401.3281  647.1078   True
Premium Very Good -602.4978  0.001  -737.2366 -467.7591   True
--------------------------------------------------------------
